File: apps/mobile-native/ios/NexaraApp/Data/SmartQuoteRepository.swift
import Foundation

/// Cotizador inteligente — paridad Android `SmartQuoteRepository`.
final class SmartQuoteRepository {
    static let shared = SmartQuoteRepository()
    private let api = ApiClient.shared
    private init() {}

    func ctProductCount() async -> Int {
        do {
            let map = ConsoleHelpers.decodeMap(try await api.get("smart-quote/ct/status"))
            return ConsoleHelpers.mapInt(map, "total")
        } catch {
            return 0
        }
    }

    func facets() async throws -> (brands: [String], categories: [String]) {
        let map = ConsoleHelpers.decodeMap(try await api.get("smart-quote/facets"))
        let brands = (map["brands"] as? [[String: Any]] ?? []).compactMap { row -> String? in
            let n = ConsoleHelpers.mapStr(row, "name")
            return n.isEmpty ? nil : n
        }
        let categories = (map["categories"] as? [[String: Any]] ?? []).compactMap { row -> String? in
            let n = ConsoleHelpers.mapStr(row, "name")
            return n.isEmpty ? nil : n
        }
        return (brands, categories)
    }

    func search(
        query: String,
        margin: Int = 30,
        optimize: String = "BALANCE",
        brand: String? = nil,
        category: String? = nil,
        limit: Int = 40
    ) async throws -> [[String: Any]] {
        var q: [String: String] = [
            "optimize": optimize,
            "targetMargin": String(margin),
            "take": String(limit),
        ]
        let trimmed = query.trimmingCharacters(in: .whitespaces)
        if !trimmed.isEmpty { q["q"] = trimmed }
        if let brand, !brand.isEmpty { q["brand"] = brand }
        if let category, !category.isEmpty { q["category"] = category }

        let data = try await api.get("smart-quote/search", query: q)
        let map = ConsoleHelpers.decodeMap(data)
        if let rows = map["data"] as? [[String: Any]] { return rows }
        return ApiClient.decodeMapList(data)
    }

    func configureTemplate(template: String, margin: Int = 30) async throws -> [[String: Any]] {
        struct Body: Encodable {
            let template: String
            let targetMarginPercent: Int
            let optimize: String
            let includeLabor: Bool
        }
        let data = try await api.postJSON(
            "smart-quote/configure",
            body: Body(template: template, targetMarginPercent: margin, optimize: "BALANCE", includeLabor: true)
        )
        let map = ConsoleHelpers.decodeMap(data)
        if let lines = map["lines"] as? [[String: Any]] { return lines }
        if let cart = map["cart"] as? [[String: Any]] { return cart }
        return []
    }

    func checkMargin(unitCost: Double, unitPrice: Double, category: String?, brand: String?) async throws -> [String: Any] {
        struct Body: Encodable {
            let unitCost: Double
            let unitPrice: Double
            let category: String?
            let brand: String?
        }
        let data = try await api.postJSON(
            "smart-quote/rules/check-margin",
            body: Body(unitCost: unitCost, unitPrice: unitPrice, category: category, brand: brand)
        )
        return ConsoleHelpers.decodeMap(data)
    }

    func substitutes(clave: String, margin: Int = 30, optimize: String = "BALANCE") async throws -> [[String: Any]] {
        let data = try await api.get(
            "smart-quote/substitutes/\(clave)",
            query: ["optimize": optimize, "targetMargin": String(margin)]
        )
        let map = ConsoleHelpers.decodeMap(data)
        if let rows = map["data"] as? [[String: Any]] { return rows }
        return ApiClient.decodeMapList(data)
    }

    func laborSuggest(cart: [[String: Any]]) async throws -> [[String: Any]] {
        struct Line: Encodable {
            let name: String
            let qty: Int
            let category: String?
        }
        struct Body: Encodable { let lines: [Line] }
        let lines: [Line] = cart.map { row in
            Line(
                name: ConsoleHelpers.mapStr(row, "nombre", "name"),
                qty: max(1, ConsoleHelpers.mapInt(row, "qty", "quantity")),
                category: ConsoleHelpers.mapStr(row, "categoria", "category", "brand").nilIfEmpty
            )
        }
        let data = try await api.postJSON("smart-quote/labor/suggest", body: Body(lines: lines))
        if let rows = ConsoleHelpers.decodeMap(data)["items"] as? [[String: Any]] { return rows }
        return ApiClient.decodeMapList(data)
    }

    func copilotDraft(prompt: String) async throws -> [String: Any] {
        struct Body: Encodable { let prompt: String }
        let data = try await api.postJSON("smart-quote/copilot/draft", body: Body(prompt: prompt))
        return ConsoleHelpers.decodeMap(data)
    }

    func logisticsZones() async throws -> [[String: Any]] {
        let data = try await api.get("smart-quote/logistics")
        return ApiClient.decodeMapList(data)
    }

    func supplierStats(from: String? = nil, to: String? = nil) async throws -> [String: Any] {
        var q: [String: String] = [:]
        if let from { q["from"] = from }
        if let to { q["to"] = to }
        let data = try await api.get("smart-quote/supplier-stats", query: q)
        return ConsoleHelpers.decodeMap(data)
    }

    func ctOrderPreview(cotizacionId: Int) async throws -> [String: Any] {
        let data = try await api.get("smart-quote/ct/orders/preview/\(cotizacionId)")
        return ConsoleHelpers.decodeMap(data)
    }

    func submitCtOrder(cotizacionId: Int, almacen: String, confirmNow: Bool = false) async throws {
        struct Body: Encodable {
            let almacen: String
            let confirmNow: Bool
        }
        _ = try await api.postJSON(
            "smart-quote/ct/orders/\(cotizacionId)",
            body: Body(almacen: almacen, confirmNow: confirmNow)
        )
    }

    func confirmCtOrder(orderId: Int) async throws {
        struct Empty: Encodable {}
        _ = try await api.postJSON("smart-quote/ct/orders/confirm/\(orderId)", body: Empty())
    }

    func createQuote(
        quoteNumber: String,
        clientName: String,
        projectName: String?,
        items: [[String: Any]]
    ) async throws -> [String: Any] {
        struct Item: Encodable {
            let name: String
            let qty: Int
            let unitPrice: Double
            let discount: Double
            let tax: Double
            let description: String?
            let category: String?
            let brand: String?
            let sku: String?
            let partNumber: String?
            let unitCost: Double?
            let marginPercent: Double?
        }
        struct Body: Encodable {
            let quoteNumber: String
            let issueDate: String
            let validUntil: String
            let clientName: String
            let projectName: String?
            let currency: String
            let status: String
            let items: [Item]
        }

        let fmt = DateFormatter()
        fmt.dateFormat = "yyyy-MM-dd"
        let today = fmt.string(from: Date())
        let valid = fmt.string(from: Calendar.current.date(byAdding: .day, value: 15, to: Date()) ?? Date())
        let year = Calendar.current.component(.year, from: Date())
        let suffix = String(Int(Date().timeIntervalSince1970) % 1_000_000)
        let generatedQuoteNumber = "NXR-\(year)-\(suffix)"
        let finalQuoteNumber = quoteNumber.isEmpty ? generatedQuoteNumber : quoteNumber

        let encItems: [Item] = items.map { row in
            let cost = ConsoleHelpers.mapDouble(row, "costMxn", "unitCost")
            let price = ConsoleHelpers.mapDouble(row, "sellPriceSuggested", "unitPrice", "precio")
            let margin = ConsoleHelpers.mapDouble(row, "marginPercent")
            return Item(
                name: ConsoleHelpers.mapStr(row, "nombre", "name"),
                qty: max(1, ConsoleHelpers.mapInt(row, "qty", "quantity")),
                unitPrice: price,
                discount: 0,
                tax: 16,
                description: ConsoleHelpers.mapStr(row, "modelo", "description").nilIfEmpty,
                category: ConsoleHelpers.mapStr(row, "categoria", "category").nilIfEmpty,
                brand: ConsoleHelpers.mapStr(row, "marca", "brand").nilIfEmpty,
                sku: ConsoleHelpers.mapStr(row, "clave", "sku").nilIfEmpty,
                partNumber: ConsoleHelpers.mapStr(row, "numParte", "partNumber").nilIfEmpty,
                unitCost: cost > 0 ? cost : nil,
                marginPercent: margin > 0 ? margin : nil
            )
        }

        let data = try await api.postJSON(
            "cotizaciones",
            body: Body(
                quoteNumber: finalQuoteNumber,
                issueDate: today,
                validUntil: valid,
                clientName: clientName,
                projectName: projectName?.nilIfEmpty,
                currency: "MXN",
                status: "DRAFT",
                items: encItems
            )
        )
        return ConsoleHelpers.decodeMap(data)
    }
}

File: apps/mobile-native/android/app/src/main/java/mx/nexara/mobile/nativeapp/data/crm/SmartQuoteRepository.kt
package mx.nexara.mobile.nativeapp.data.crm

import android.content.Context
import mx.nexara.mobile.nativeapp.data.AuthRepository
import mx.nexara.mobile.nativeapp.data.api.ApiClient
import mx.nexara.mobile.nativeapp.data.api.ConfigureSolutionResponse
import mx.nexara.mobile.nativeapp.data.api.CopilotDraftDto
import mx.nexara.mobile.nativeapp.data.api.CreateCotizacionBody
import mx.nexara.mobile.nativeapp.data.api.CreateCotizacionItemBody
import mx.nexara.mobile.nativeapp.data.api.CtOrderPreviewDto
import mx.nexara.mobile.nativeapp.data.api.LaborSuggestBody
import mx.nexara.mobile.nativeapp.data.api.LaborSuggestItemDto
import mx.nexara.mobile.nativeapp.data.api.CommercialRuleDto
import mx.nexara.mobile.nativeapp.data.api.LogisticsZoneDto
import mx.nexara.mobile.nativeapp.data.api.MarginCheckBody
import mx.nexara.mobile.nativeapp.data.api.MarginCheckDto
import mx.nexara.mobile.nativeapp.data.api.SmartFacetsResponseDto
import mx.nexara.mobile.nativeapp.data.api.SmartOfferDto
import mx.nexara.mobile.nativeapp.data.api.SmartQuoteApi
import mx.nexara.mobile.nativeapp.data.api.SupplierStatsResponseDto

class SmartQuoteRepository(context: Context) {
    private val api: SmartQuoteApi =
        ApiClient.authed { AuthRepository(context).token() }.create(SmartQuoteApi::class.java)

    suspend fun search(
        query: String,
        margin: Int = 30,
        optimize: String = "BALANCE",
        brand: String? = null,
        category: String? = null,
    ): List<SmartOfferDto> {
        if (query.isBlank() && brand.isNullOrBlank() && category.isNullOrBlank()) return emptyList()
        return api.search(
            q = query.trim().ifBlank { null },
            brand = brand?.takeIf { it.isNotBlank() },
            category = category?.takeIf { it.isNotBlank() },
            optimize = optimize,
            targetMargin = margin,
            take = 40,
        ).data
    }

    suspend fun facets(): SmartFacetsResponseDto = api.facets()

    suspend fun substitutes(clave: String, margin: Int = 30, optimize: String = "BALANCE"): List<SmartOfferDto> =
        api.substitutes(clave, optimize, margin)

    suspend fun laborSuggest(cart: List<QuoteCartLine>): List<LaborSuggestItemDto> {
        val lines = cart.map {
            mapOf(
                "name" to it.name,
                "qty" to it.qty,
                "category" to (it.brand ?: "CT"),
            )
        }
        return api.laborSuggest(LaborSuggestBody(lines))
    }

    suspend fun marginRules(): List<CommercialRuleDto> = runCatching { api.marginRules() }.getOrDefault(emptyList())

    suspend fun logisticsZones(): List<LogisticsZoneDto> = runCatching { api.logisticsZones() }.getOrDefault(emptyList())

    suspend fun checkMargin(
        unitCost: Double,
        unitPrice: Double,
        category: String? = null,
        brand: String? = null,
    ): MarginCheckDto =
        api.checkMargin(
            MarginCheckBody(
                unitCost = unitCost,
                unitPrice = unitPrice,
                category = category,
                brand = brand,
            ),
        )

    suspend fun copilotDraft(prompt: String): CopilotDraftDto =
        api.copilotDraft(mx.nexara.mobile.nativeapp.data.api.CopilotDraftBody(prompt))

    suspend fun configureSolution(
        template: String,
        cameras: Int? = null,
        storageDays: Int? = null,
        accessPoints: Int? = null,
        doors: Int? = null,
        margin: Int = 30,
        optimize: String = "BALANCE",
        logisticsZone: String? = null,
    ): ConfigureSolutionResponse =
        api.configure(
            mx.nexara.mobile.nativeapp.data.api.ConfigureSolutionBody(
                template = template,
                cameras = cameras,
                storageDays = storageDays,
                accessPoints = accessPoints,
                doors = doors,
                optimize = optimize,
                targetMarginPercent = margin,
                logisticsZone = logisticsZone,
                includeLabor = true,
            ),
        )

    suspend fun ctProductCount(): Int = runCatching { api.ctStatus().total }.getOrDefault(0)

    suspend fun supplierStats(from: String? = null, to: String? = null): SupplierStatsResponseDto =
        api.supplierStats(from = from, to = to)

    suspend fun ctOrderPreview(cotizacionId: Long): CtOrderPreviewDto =
        api.ctOrderPreview(cotizacionId)

    suspend fun submitCtOrder(cotizacionId: Long, almacen: String, confirmNow: Boolean = false) {
        api.ctOrderSubmit(cotizacionId, mx.nexara.mobile.nativeapp.data.api.CtOrderSubmitBody(almacen, confirmNow))
    }

    suspend fun confirmCtOrder(orderId: Long) {
        api.ctOrderConfirm(orderId)
    }

    suspend fun createQuote(
        clientName: String,
        projectName: String?,
        lines: List<QuoteCartLine>,
        optimizeMode: String = "BALANCE",
    ): Long {
        val now = java.time.LocalDate.now()
        val quoteNumber = "NXR-${now.year}-${System.currentTimeMillis().toString().takeLast(6)}"
        val body = CreateCotizacionBody(
            quoteNumber = quoteNumber,
            issueDate = now.toString(),
            validUntil = now.plusDays(15).toString(),
            clientName = clientName,
            clientCompany = clientName,
            projectName = projectName?.takeIf { it.isNotBlank() },
            items = lines.map { l ->
                CreateCotizacionItemBody(
                    name = l.name,
                    qty = l.qty,
                    unitPrice = l.unitPrice,
                    discount = 0.0,
                    tax = 16.0,
                    category = if (l.isLabor) "LABOR" else l.brand,
                    brand = l.brand,
                    model = l.model,
                    sku = l.sku,
                    partNumber = l.partNumber,
                    unitCost = l.unitCost,
                    supplierCode = if (l.productCtId != null) "CT" else l.supplierCode,
                    productCtId = l.productCtId,
                    supplierSku = l.sku,
                    marginPercent = l.marginPercent,
                    stockSnapshot = l.stockSnapshot,
                    optimizationMode = optimizeMode,
                )
            },
        )
        val res = api.createCotizacion(body)
        return res.id ?: throw IllegalStateException("Cotización sin ID")
    }
}

data class QuoteCartLine(
    val productCtId: Long?,
    val name: String,
    val category: String? = null,
    val brand: String?,
    val model: String?,
    val sku: String?,
    val partNumber: String?,
    val qty: Int,
    val unitCost: Double,
    val unitPrice: Double,
    val marginPercent: Double,
    val stockSnapshot: Int?,
    val isLabor: Boolean = false,
    val supplierCode: String? = null,
)
